app/src/org/commcare/preferences/HiddenPreferences.java

```diff
@@ -320,4 +325,30 @@ public static boolean shouldShowUnsentFormsWhenZero() {
         SharedPreferences properties = CommCareApplication.instance().getCurrentApp().getAppPreferences();
         return properties.getString(SHOW_UNSENT_FORMS_WHEN_ZERO, PrefValues.NO).equals(PrefValues.YES);
     }
+
+    public static boolean preUpdateSyncNeeded() {
+        SharedPreferences properties = CommCareApplication.instance().getCurrentApp().getAppPreferences();
+        return properties.getString(PRE_UPDATE_SYNC_NEEDED, PrefValues.NO).equals(PrefValues.YES);
+    }
+
+    public static void setReleasedOnTimeForOngoingAppDownload(AndroidCommCarePlatform platform, long releasedOnTime) {
+        if (platform.getApp() == null) {
+            return;
+        }
+
+        platform.getApp().getAppPreferences()
+                .edit()
+                .putLong(RELEASED_ON_TIME_FOR_ONGOING_APP_DOWNLOAD, releasedOnTime)
+                .apply();
+    }
+
+    public static long geReleasedOnTimeForOngoingAppDownload() {
+        if (CommCareApplication.instance().getCurrentApp() == null) {
+            return 0;
+        }
+
+        return CommCareApplication.instance().getCurrentApp().getAppPreferences()
+                .getLong(RELEASED_ON_TIME_FOR_ONGOING_APP_DOWNLOAD, 0);
```

Comment:
I think we need a "sanity check" for this date/time. Once it's set, it is set pretty much forever and if, say, the phone had its date set wrong when the user downloaded the update (say, to the future), then failed to sync and then set the phone's time back to the correct time, this would constantly block them from updating.

Since the phone's date can change, I think we can only do that sanity checking when the value is *read* not when it is *written*, so it would seem to make sense to me to do it here.